src/expr/type.cpp

```diff
@@ -566,7 +569,14 @@ std::vector<Type> ConstructorType::getArgTypes() const {
 
 const Datatype& DatatypeType::getDatatype() const {
   NodeManagerScope nms(d_nodeManager);
-  return d_typeNode->getDatatype();
+  Assert(isDatatype());
+  if (d_typeNode->getKind() == kind::DATATYPE_TYPE)
+  {
+    DatatypeIndexConstant dic = d_typeNode->getConst<DatatypeIndexConstant>();
+    return d_nodeManager->getDatatypeForIndex(dic.getIndex());
+  }
+  Assert(d_typeNode->getKind() == kind::PARAMETRIC_DATATYPE);
+  return static_cast<DatatypeType>((*d_typeNode)[0].toType()).getDatatype();
```

Comment:
Strictly speaking, this is undefined behavior because `(*d_typeNode)[0].toType()` is not of type `DatatypeType` (UBSan should complain about this). The correct way is:

```suggestion
  return DatatypeType((*d_typeNode)[0].toType()).getDatatype();
```